gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageGrpcIntegrationTest.java

```diff
@@ -37,31 +44,58 @@ public class GoogleCloudStorageGrpcIntegrationTest {
 
   private static final String BUCKET_NAME = BUCKET_HELPER.getUniqueBucketName("shared");
 
-  private static GoogleCloudStorage createGoogleCloudStorage() throws IOException {
+  private final boolean tdEnabled;
+
+  @Parameters
+  // We want to test this entire class with both gRPC-LB and Traffic Director
+  public static Iterable<Boolean> tdEnabled() {
+    return Arrays.asList(false, true);
+  }
+
+  public GoogleCloudStorageGrpcIntegrationTest(boolean tdEnabled) {
+    this.tdEnabled = tdEnabled;
+  }
+
+  private GoogleCloudStorageOptions.Builder configureOptions() {
+    GoogleCloudStorageOptions.Builder optionsBuilder =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setGrpcEnabled(true)
+            .setTrafficDirectorEnabled(this.tdEnabled);
+    boolean runningTestsFromGCP = Boolean.parseBoolean(System.getenv("RUNNING_TESTS_FROM_GCP_VM"));
+    if (this.tdEnabled && runningTestsFromGCP) {
+      optionsBuilder.setGrpcServerAddress("storage-force-dp.googleusercontent.com");
+      logger.atInfo().log("Overriding gRPC server address to force-dp endpoint");
+    }
+    return optionsBuilder;
+  }
+
+  private GoogleCloudStorage createGoogleCloudStorage() throws IOException {
     return new GoogleCloudStorageImpl(
-        GoogleCloudStorageTestHelper.getStandardOptionBuilder().setGrpcEnabled(true).build(),
-        GoogleCloudStorageTestHelper.getCredentials());
+        configureOptions().build(), GoogleCloudStorageTestHelper.getCredentials());
   }
 
-  private static GoogleCloudStorage createGoogleCloudStorage(
+  private GoogleCloudStorage createGoogleCloudStorage(
       AsyncWriteChannelOptions asyncWriteChannelOptions) throws IOException {
     return new GoogleCloudStorageImpl(
-        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
-            .setWriteChannelOptions(asyncWriteChannelOptions)
-            .setGrpcEnabled(true)
-            .build(),
+        configureOptions().setWriteChannelOptions(asyncWriteChannelOptions).build(),
         GoogleCloudStorageTestHelper.getCredentials());
   }
 
   @BeforeClass
   public static void createBuckets() throws IOException {
-    GoogleCloudStorage rawStorage = createGoogleCloudStorage();
+    GoogleCloudStorage rawStorage =
+        new GoogleCloudStorageImpl(
+            GoogleCloudStorageTestHelper.getStandardOptionBuilder().setGrpcEnabled(true).build(),
+            GoogleCloudStorageTestHelper.getCredentials());
     rawStorage.createBucket(BUCKET_NAME);
   }
 
   @AfterClass
   public static void cleanupBuckets() throws IOException {
-    GoogleCloudStorage rawStorage = createGoogleCloudStorage();
+    GoogleCloudStorage rawStorage =
+        new GoogleCloudStorageImpl(
+            GoogleCloudStorageTestHelper.getStandardOptionBuilder().setGrpcEnabled(true).build(),
+            GoogleCloudStorageTestHelper.getCredentials());
```

Comment:
Do we need this change ? Let's reuse the changes in the `createGoogleCloudStorage` method

Reply:
Yes, these are needed because @BeforeClass method has to be static while `createGoogleCloudStorage` is no longer static because of using the class fields.

Reply:
Refactored this now to have common endpoint logic for everyone. So when we want to run tests on pre-prod for example buckets will also be created in same universe.